solr/solrj/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -326,12 +408,11 @@ public void forciblyRefreshAllClusterStateSlow() throws KeeperException, Interru
       // No need to set watchers because we should already have watchers registered for everything.
       refreshCollectionList(null);
       refreshLiveNodes(null);
-      // Need a copy so we don't delete from what we're iterating over.
-      Collection<String> safeCopy = new ArrayList<>(watchedCollectionStates.keySet());
+
       Set<String> updatedCollections = new HashSet<>();
-      for (String coll : safeCopy) {
+      for (String coll : collectionWatches.keySet()) {
```

Comment:
I don't think that we actually want any of the null collections, right?

They would only be null if they stopped existing, correct? I might be wrong here

Reply:
Ah yes. This is one of the change that Im not 100% sure as it changes the behavior. It was buried in one of the original questions:

1. In `forciblyRefreshAllClusterStateSlow`, which we are proposing to just replace `watchedCollectionStates.keySet()` with `collectionWatches.keySet()`, the reasoning is that we should forcibly update every collection that is registered in `collectionWatches`, even if previous fetch on such collection returned null (ie `watchedCollectionStates` would have no entry on it). This is a minor change of behavior but I think it is more "correct"? Would really need more experienced Solr dev to share their thoughts here 🙏🏼